### DISCUSSION TITLE
How to get the real IP address in the clients

### DISCUSSION THREAD

**Discussion by Waldorf3** [NginxProxyManager/nginx-proxy-manager#3215] [Q&A]
So my setup is as follows.

I have a Proxmox server with a number of virtual machines. 

All virtual macines are interconnected with private IP addresses (like 10.10.10.x, with x essentially being the VMID)

One virtual machine additionally has a public ip address.  It runs nginx-proxy-manager in docker, and receives the traffic on the public IP address, and forwards the traffice to the other virtual machines through the private IP addresses.

This works perfectly except the virtual machines all see the traffic (the IP) as coming from the nginx-proxy-manager.

I've spent hours now trying to get the real client IP address to appear in the access log files on the proxy clients, but nothing works.

So I tried leaving the customer nginx configuration (in the gui) empty, and I have tried adding the lines

```
set_real_ip_from  IP;
real_ip_header X-Forwarded-For;
```

With IP being both the public IP address and the private IP address.  I tried other combinations I found here and elsewhere, but the bottom line is, nothing ever changes.

On the clients I also tried adding

```
                RemoteIPHeader X-Forwarded-For
                RemoteIPInternalProxy IP
```

Again with IP being both the private and the public IP on the nginx-proxy-server, but nothing changes anything.

Surely there must be one correct way to set this up?  Is this documented somewhere? I haven't been able to find any, other than man discussions here, and that isn't always clear and usually different situations.

**Comment by gillescoolen** [comment]
Did you make any progress on the issue? Currently having the same problem, and the fix in [this issue](https://github.com/NginxProxyManager/nginx-proxy-manager/issues/112) does nothing for me.

**Comment by pachylover** [comment]
In my case, the issue was caused by trying to get X-Forwarded-For when it was not HTTPS.

**Comment by BuzzMoody** [comment]
Anyone got a fix for this?

**Comment by Crash1602** [comment]
My question is relevant to the topic, so I'm posting it here.

How can I forward the source IP for a TCP stream? Currently only the Docker Host IP is displayed on the upstream server, so i can´t use Autoblock-IP by failed logins.

Thank you very much.

**Reply by baskinsy** [reply]
I am looking for this also and can't find something.

**Comment by begunfx** [comment]
**Synology workaround fix for this!**

I found a solution for this if you are trying to run NPM (Docker) on a Synology.  It seems that there are some pre-routing rules that need to be added to the Synology host for the IP addresses to report the client and not Docker.  

I found this post with the solution:

> https://www.pedrolamas.com/2020/11/04/exposing-the-client-ips-to-docker-containers-on-synology-nas/

The short of it is you need to apply the following iptable rules on Synology (cli):
`sudo iptables -t nat -A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER`
`sudo iptables -t nat -A PREROUTING -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER`

Only catch is that these changes will not sustain post reboot. 

In order for that to happen you need to create a boot-up task in the task scheduler and add the script he created so that the rule changes are applied every startup.

I tested the rule changes and it seems I can now properly use the Access List and restrict access to local (LAN) only.
script and instructions here:

> https://gist.github.com/pedrolamas/db809a2b9112166da4a2dbf8e3a72ae9

**Reply by ghost** [reply]
What is the effect of this rule on the firewall?

**Comment by cowpiebrownrice** [comment]
Hi @Waldorf3!

> This works perfectly except the virtual machines all see the traffic (the IP) as coming from the nginx-proxy-manager.

I had the same problem.
After many researches and tests, this solution works for me:
Adding:
`proxy_set_header X-Real-IP $remote_addr;`
in Nginx Proxy Manager / Edit Proxy Host / Advanced / Custom Nginx Configuration.

I hope it can be useful. :+1:

**Reply by filmo** [reply]
Hi @rardcode

In the Advance field, did you add anything other than just exactly: `proxy_set_header X-Real-IP $remote_addr;`
such as any other brackets or config details??

When I added just "`proxy_set_header X-Real-IP $remote_addr;`" and nothing else to my Advance tab, I'm still seeing X-Real-IP show up as my the IP address of my router rather than the external IP address.

My set up is: External request => Router with Port Forwarding => NPM => Webserver
By the time my webserver gets the request 'REMOTE_ADDR' is set to the IP address of the NPM and HTTP_X_REAL_IP is set to the IP address of my router. And the external IP request is 'lost'

**Reply by cowpiebrownrice** [reply]
Hi @filmo 
in npm config i have:
```
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $remote_addr;
```
I use NPM before Nextcloud server. In Nextcloud `config.php` i have added:
```
  'forwarded_for_headers' =>
  array (
    0 => 'HTTP_X_FORWARDED_FOR',
  ),
```
I hope this can help you!

**Reply by segobytes** [reply]
Hi @filmo , I have the same issue: I only get the gateway IP address while using Synology’s nginx. I’m wondering if I should set NPM in a docker instead. Did you find a solution ?

**Reply by walterzilla** [reply]
> Hi @filmo , I have the same issue: I only get the gateway IP address while using Synology’s nginx. I’m wondering if I should set NPM in a docker instead. Did you find a solution ?

Did you try setting ­`trusted_proxies` in Nextcloud config.php also?

**Reply by segobytes** [reply]
I don't use nextcloud, I use a synology, but I did set the trusted proxies, and no change. Basically without any additional settings, all the clients connecting from outside show up with the nginx reverse proxy host internal ip (192.168.0.9) and if I fill the trusted proxies with;
- either just 192.168.0.9: the external clients show up as 192.168.0.1 (my gateway)
- or 192.168.0.9 and 192.168.0.1: the external clients show up as 192.168.0.1 (my gateway)
At that stage I gave up trying to make my current config work (synology integrated reverse proxy feature), and I'd rather implement a solution known to work, hence my question about setting up NPM in a separate docker: would that work or would I end up in the samle situation ....

**Reply by walterzilla** [reply]
> I don't use nextcloud

Sorry dude yesterday was seeking hard for a solution for Nextcloud and well I got distracted so to speak......

**Comment by Yusseiin** [comment]
Someone has been able to fix this?
Nothing in the "Custom Nginx Configuration" seems to working, always getting back the Docker IP

I am on Ubuntu Server

**Comment by alemstrom** [comment]
probably the most useless post here

**Comment by segobytes** [comment]
Yes, the problem was at my router settings level; I had forgotten that I had set up the following:
iptables -t nat -A POSTROUTING -j MASQUERADE
...and this is what was creating the issue: MASQUERADE rewrites the source IP of outgoing packets to the router’s own WAN IP.
and this is why my Synology reverse proxy only saw the router’s IP instead of the real client IP.
so I had to remove that rule with: 
iptables -t nat -D POSTROUTING -j MASQUERADE  # Remove existing rule
then I added the following rules
# Allow LAN traffic without NAT (Keep original source IP)
iptables -t nat -A POSTROUTING -p tcp -d 192.168.0.0/16 -j RETURN

# Keep real client IPs for HTTPS traffic
iptables -t nat -A POSTROUTING -p tcp -d <synology_ip> --dport 443 -j RETURN

# Masquerade everything else (for outbound internet traffic)
iptables -t nat -A POSTROUTING -j MASQUERADE
and now I it works as intended: the reverse proxy can see if it is a an external ip and so I can define different behaviours depensing on internal or external client.

**Comment by pietabrow** [comment]
Fixed it with this:
Open your Nginx Proxy Manager web portal
Go to Hosts > Proxy Hosts > Edit > Advanced > Custom Nginx Configuration
add the following:
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $remote_addr;
Click the Save button

Now login to your Nextcloud server commandline via SSH
Edit your Nextcloud config.php file, like this:
sudo nano /var/www/nextcloud/config/config.php
(maybe your path is slightly different)
add the following:
  'trusted_proxies' => ['192.168.1.150'],
  'forwarded_for_headers' => ['HTTP_X_FORWARDED', 'HTTP_FORWARDED_FOR'],
(make sure you change 192.168.1.150 to the correct IP address from your Nginx Proxy Manager)
Save and exit (ctrl+x, Y, enter)
Now the correct IP addresses should be presented to your Nextcloud server.

**Comment by DerGeraet69** [comment]
I found a solution for me with NPM and Nextcloud. 
Setting it in the Web-Gui isn't supported!!

[Reverse Proxy](https://docs.nextcloud.com/server/31/admin_manual/configuration_server/reverse_proxy_configuration.html) says "Nextcloud uses the de-facto standard header ‘X-Forwarded-For’ by default", so setting the 'forwarded_for_headers' isn't needed. 

Under ...data/nginx/proxy_host NPM stores the configuration for the Hosts you configure in the webinterface and i noticed following in the location section in my "13.conf" for Nextcloud (for you it might be a different number).

``` 
location / {

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;

    # Proxy!
    include conf.d/include/proxy.conf;
  }
```
There is already a include for a folder conf.d and a proxy.conf, which isn't created yet!
I added "...data/nginx/proxy_host/conf.d/proxy.conf" with following content:

```
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
```
Since all standard conf for the proxy hosts are similar this should work for every other hosts, but i haven't confirmed yet.

Then i just had to add the trusted_proxies to the ".../www/nextcloud/config/config.php".
Maybe a restart of the Proxy is needed, but after that i got the real IP in the nextcloud.log for failed logins. 

I wanted to add Fail2Ban to nextcloud but i noticed that it already as a own "Brute-Force" module, which limits attempts for an IP.

Setting the network_mode to host in the docker-compose.yaml wasn't needed for me.

**Comment by fvultee** [comment]
I've tried this exactly solution and restarted NPM, made the folder, made
the proxy.conf file, etc, I can see it's called during NPM reboot in the
logs, still nothing.  There isn't a place in Bitwarden (where I'm currently
trying to get the real IP) to specify trusted proxies, only trusted IPs,
which I've done.  Still grabs the Container IP.

**Reply by DerGeraet69** [reply]
I can only guess why it works for me., even long before i made the changes for Nextcloud. I'm using Vaultwarden and in the admin panel i noticed under advanced settings -> Client IP Header -> "X-Real-IP". Tried to verify this by removing the inlude from the proxy-host config and got a error "KDF config is required" which doesn't seem to be related. Maybe somone can explain. After changing back it worked again.

**Reply by fvultee** [reply]
It randomly started to work for me, so strange.  I blew away my containers and started again.  Now I get the real IP in Bitwarden from Nginix Proxy Manager and I can ban that IP via fail2ban.  Now I'm trying to get fail2ban emails working, which hasn't been as easy as it should be!  always somethin!

**Reply by fvultee** [reply]
Well, gave up on emails working, but I got block / unblock notifications to push to discord so that's great!  All is finally working.

**Comment by THEDARKK** [comment]
Hello,
I'm having the same issue.
I'm using a dedicated server, and I always get the local IP address of the proxy instead of the actual IP address.

**Reply by borekon** [reply]
https://github.com/NginxProxyManager/nginx-proxy-manager/discussions/3215#discussioncomment-13855070
https://github.com/NginxProxyManager/nginx-proxy-manager/discussions/3215#discussioncomment-13404336

just read a bit

**Comment by AChalukov** [comment]
In my case, on TrueNAS Scale version 25.04.2.6, I had to create a custom Nginx configuration file inside the Nextcloud config dataset and then point to it from the Nginx container. This was obviously a mistake on my part, but maybe someone else has made it too.

Here is the content of the custom Nginx file (you can change it to fit your Docker network if it is different):


<img width="512" height="321" alt="Screenshot 2025-12-09 021117" src="https://github.com/user-attachments/assets/660a0121-be13-4d84-81e1-e4bb1642ddd6" />

nginx_custome.conf:

set_real_ip_from  172.16.0.0/12;
set_real_ip_from  192.168.0.0/16;
set_real_ip_from  10.0.0.0/8;

real_ip_header    X-Forwarded-For;
real_ip_recursive on;


Make sure your NAT rules are also correct. I had a problem with my previous hairpin NAT rule, which just slowed me down while analyzing the issue.

In the Nextcloud container logs, I could still see the Docker IP along with the public client IP (tested with Tor Browser and mobile data).

Also, I did not need to add any Custom Nginx Configuration in Nginx Proxy Manager.

**Comment by rogercreagh** [comment]
I was tearing my remaining hair out because nothing suggested above seemed to work. Then I happened to notice that my connection was using an ip6 address. Clients connecting with ip4 were getting HTTP_X_REAL_IP set correctly, but when using ip6 the internal address was appearing.

Quick fix was to remove the ip6 address from the server..

Have tried following [issue 4029](https://github.com/NginxProxyManager/nginx-proxy-manager/issues/4029) but simple adding `proxy_set_header Host $host; ` in the settings for the host in NPM didn't seem to work with NPM v2.15.1 didn't seem to work.

Any ideas?
